idea-plugin/src/main/java/com/palantir/javaformat/intellij/PalantirJavaFormatFormattingService.java

```diff
@@ -0,0 +1,157 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.javaformat.intellij;
+
+import static java.util.Comparator.comparing;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Range;
+import com.intellij.formatting.service.AsyncDocumentFormattingService;
+import com.intellij.formatting.service.AsyncFormattingRequest;
+import com.intellij.ide.highlighter.JavaFileType;
+import com.intellij.lang.ImportOptimizer;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.NlsSafe;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiFile;
+import com.palantir.javaformat.java.FormatterException;
+import com.palantir.javaformat.java.FormatterService;
+import com.palantir.javaformat.java.Replacement;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+
+class PalantirJavaFormatFormattingService extends AsyncDocumentFormattingService {
+    private final FormatterProvider formatterProvider = new FormatterProvider();
+
+    @Override
+    protected FormattingTask createFormattingTask(@NotNull AsyncFormattingRequest request) {
+        Project project = request.getContext().getProject();
+        PalantirJavaFormatSettings settings = PalantirJavaFormatSettings.getInstance(project);
+        Optional<FormatterService> formatter = formatterProvider.get(project, settings);
+        return new PalantirJavaFormatFormattingTask(request, formatter);
+    }
+
+    @Override
+    protected @NotNull String getNotificationGroupId() {
+        return Notifications.PARSING_ERROR_NOTIFICATION_GROUP;
+    }
+
+    @Override
+    protected @NotNull @NlsSafe String getName() {
+        return "palantir-java-format";
+    }
+
+    @Override
+    public @NotNull Set<Feature> getFeatures() {
+        return Set.of(Feature.AD_HOC_FORMATTING, Feature.FORMAT_FRAGMENTS, Feature.OPTIMIZE_IMPORTS);
+    }
+
+    @Override
+    public boolean canFormat(@NotNull PsiFile file) {
+        return JavaFileType.INSTANCE.equals(file.getFileType())
+                && PalantirJavaFormatSettings.getInstance(file.getProject()).isEnabled();
+    }
+
+    @Override
+    public @NotNull Set<ImportOptimizer> getImportOptimizers(@NotNull PsiFile file) {
+        Project project = file.getProject();
+        PalantirJavaFormatSettings settings = PalantirJavaFormatSettings.getInstance(project);
+        Optional<FormatterService> formatter = formatterProvider.get(project, settings);
+        return Set.of(new PalantirJavaFormatImportOptimizer(formatter));
```

Comment:
different from google-java-format: we need to pass the formatter to the Optimizer class, such that we can call the spi implementation